share/completions/git.fish

```diff
@@ -737,16 +661,39 @@ function __fish_git_complete_stashes
     __fish_git stash list --format=%gd:%gs 2>/dev/null | string replace ":" \t
 end
 
-function __fish_git_aliases
-    __fish_git config -z --get-regexp '^alias\.' 2>/dev/null | while read -lz key value
-        begin
-            set -l name (string replace -r '^.*\.' '' -- $key)
-            set -l val (string shorten --no-newline -m 36 -- $value)
-            printf "%s\t%s\n" $name "alias: $val"
+function __fish_git_get_aliases
+    # Get valid aliases - aliases for a builtin or custom commands are ignored
+    set -l non_aliases (__fish_git --list-cmds=main,others)
+    __fish_git config --get-regexp '^alias\.' 2>/dev/null | while read -l key value
+        set -l name (string replace -r '^alias\.' '' -- $key)
+        if not contains -- $name $non_aliases
+            printf "%s\t%s\n" $name $value
         end
     end
 end
 
+function __fish_git_complete_aliases
+    # List the defined aliases when completing a command
+    __fish_git_get_aliases | while read -l -d\t name value
+        printf "%s\talias: %s\n" $name (string shorten --no-newline -m 36 -- $value)
+    end
+end
+
+function __fish_git_using_alias
+    # Check whether the current completion is for an alias
+    set -l cmd (__fish_git_needs_command)
+    test -n "$cmd"
+    and contains -- "$cmd" (__fish_git_get_aliases | string split -f 1 \t)
+end
+
+function __fish_git_complete_for_alias
+    # Resolve the alias and complete the command
+    # Assumes that __fish_git_using_alias is true
+    set -l alias (__fish_git_needs_command)
+    set -l value (git config --get "alias.$alias")
+    complete -C (commandline -p | string replace -- "$alias" "$value")
```

Comment:
Oh, also: This will replace the first occurence of whatever characters.

That means if you have an alias "it"... it will replace the "it" in "git". The logic needs to be a good deal more complex:

- Find the subcommand
- Replace *that* with the alias value
- Run *that* via `complete -C`

Reply:
You can also see that the tests are failing - this doesn't complete an alias of `re` for `restore --staged` properly, there is some difference.